File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def check(self, weights, capacity, days):
        currDays, load = 1, 0
        for weight in weights:
            if weight > capacity: 
                return False

            if load + weight <= capacity:
                load += weight
            
            else:
                currDays += 1
                load = weight

        return currDays <= days

    def shipWithinDays(self, weights, days):
        """
        :type weights: List[int]
        :type days: int
        :rtype: int
        """
        sum = 0
        for weight in weights:
            sum += weight
        
        low, high, ans = 0, sum, -1

        while low <= high:
            mid = low + (high - low) // 2
            res = self.check(weights, mid, days)
            if(res):
                ans = mid
                high = mid - 1
            else:
                low = mid + 1
        
        return ans

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int sum = 0;
		for(int i = 0;i < weights.size();i++) 
			sum += weights[i];

		int low = 0, high = sum, ans = -1;
		while(low <= high) {
			int mid = low + (high - low) / 2;
			if (check(weights, mid, days)) {
			    ans = mid;
			    high = mid - 1;
			} else low = mid + 1;
		}
		return ans;
    }

private:
    bool check(vector<int>& weights, int capacity, int days) {
        int currDays = 1;
        int load = 0;
        for (int weight : weights) {
            if (weight > capacity) return false;

            if (load + weight <= capacity) load += weight;
            
            else {
                currDays++;
                load = weight;
            }
        }
        return currDays <= days;
    }
};